File: frontend/ChatApp.Frontend/src/services/authService.ts
import { authAxios } from "./axiosInstance";
import { API_ROUTES } from "../constants/routes";

export interface LoginRequest {
  username: string;
  password: string;
}

export interface RegisterRequest {
  username: string;
  password: string;
}

interface BackendAuthResponse {
  id: string;
  token: string;
  username: string;
  expiresAt: string;
}

export interface AuthResponse {
  user: {
    id: string;
    username: string;
  };
  token: string;
}

export const authService = {
  login: async (credentials: LoginRequest): Promise<AuthResponse> => {
    const response = await authAxios.post<BackendAuthResponse>(
      API_ROUTES.AUTH.LOGIN,
      credentials,
    );

    return {
      token: response.data.token,
      user: {
        id: response.data.id,
        username: response.data.username,
      },
    };
  },

  register: async (userData: RegisterRequest): Promise<AuthResponse> => {
    const response = await authAxios.post<BackendAuthResponse>(
      API_ROUTES.AUTH.REGISTER,
      userData,
    );

    return {
      token: response.data.token,
      user: {
        id: response.data.id,
        username: response.data.username,
      },
    };
  },

  logout: async (): Promise<void> => {
    await authAxios.post(API_ROUTES.AUTH.LOGOUT);
  },
};

File: backend/ChatApp.Auth/src/Application/Services/AuthService.cs
using Application.DTOs;
using ChatApp.Auth.Domain.Entities;
using Domain.Common.Constants;
using Microsoft.AspNetCore.Identity;
using Microsoft.Extensions.Options;
using Application.Interfaces;
using Domain.Settings;

namespace Application.Services;

public class AuthService : IAuthService
{
    private readonly UserManager<ApplicationUser> _userManager;
    private readonly IJwtTokenGenerator _jwtTokenGenerator;
    private readonly JwtSettings _jwtSettings;

    public AuthService(
        UserManager<ApplicationUser> userManager,
        IJwtTokenGenerator jwtTokenGenerator,
        IOptions<JwtSettings> jwtSettings)
    {
        _userManager = userManager;
        _jwtTokenGenerator = jwtTokenGenerator;
        _jwtSettings = jwtSettings.Value;
    }

    public async Task<AuthResponse> RegisterAsync(RegisterRequest request, CancellationToken cancellationToken = default)
    {
        var existingUser = await _userManager.FindByNameAsync(request.Username);
        if (existingUser != null)
        {
            throw new InvalidOperationException("Username already exists");
        }

        var user = new ApplicationUser
        {
            UserName = request.Username,
            CreatedAt = DateTime.UtcNow,
            IsActive = true
        };

        var result = await _userManager.CreateAsync(user, request.Password);
        if (!result.Succeeded)
        {
            throw new InvalidOperationException(string.Join(", ", result.Errors.Select(e => e.Description)));
        }

        await _userManager.AddToRoleAsync(user, RoleConstants.User);

        var token = await _jwtTokenGenerator.GenerateTokenAsync(user);

        return new AuthResponse(
            user.Id,
            token,
            user.UserName!,
            DateTime.UtcNow.AddMinutes(_jwtSettings.ExpirationInMinutes)
        );
    }

    public async Task<AuthResponse> LoginAsync(LoginRequest request, CancellationToken cancellationToken = default)
    {
        var user = await _userManager.FindByNameAsync(request.Username);
        if (user == null || !user.IsActive)
        {
            throw new UnauthorizedAccessException("Invalid credentials");
        }

        var isPasswordValid = await _userManager.CheckPasswordAsync(user, request.Password);
        if (!isPasswordValid)
        {
            throw new UnauthorizedAccessException("Invalid credentials");
        }

        var token = await _jwtTokenGenerator.GenerateTokenAsync(user);

        return new AuthResponse(
            user.Id,
            token,
            user.UserName!,
            DateTime.UtcNow.AddMinutes(_jwtSettings.ExpirationInMinutes)
        );
    }

    public async Task<bool> ValidateTokenAsync(string token, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(token))
            return false;

        var tokenHandler = new System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler();
        
        try
        {
            var key = System.Text.Encoding.UTF8.GetBytes(_jwtSettings.Secret);
            tokenHandler.ValidateToken(token, new Microsoft.IdentityModel.Tokens.TokenValidationParameters
            {
                ValidateIssuerSigningKey = true,
                IssuerSigningKey = new Microsoft.IdentityModel.Tokens.SymmetricSecurityKey(key),
                ValidateIssuer = true,
                ValidIssuer = _jwtSettings.Issuer,
                ValidateAudience = true,
                ValidAudience = _jwtSettings.Audience,
                ValidateLifetime = true,
                ClockSkew = TimeSpan.Zero
            }, out _);
            
            return true;
        }
        catch
        {
            return false;
        }
    }
}